### PR TITLE
HDDS-5067. [FSO] Add documentation

### PR DESCRIPTION
**Pull request by rakeshadr**
## What changes were proposed in this pull request?

This task is to do feature documentation page.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5067

## How was this patch tested?

NA

### CODE REVIEW

**Review by mukul1987 — APPROVED** [review]
+1, LGTM

**Review by elek — COMMENTED** [review]
Thanks to create this patch @prashantpogde 

I really like the quick summary, I think this is the important part: explaining what are the benefits of turning of this feature and what are the consequences.

I also like the summary presentation uploaded to the parent jira. I would suggest including the main diagram in the doc, which explains the differences very well:

image

But the two other (delete / rename) pictures also can be included (IMHO) as they explain the problem very well.

File: hadoop-hdds/docs/content/feature/PrefixFSO.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+
+The prefix based FileSystem optimization feature supports atomic rename and
+ delete of any directory at any level in the namespace. Also, it will perform
+  rename and delete of any directory in a deterministic/constant time.
+
+
```
Comment (elek) [inline hadoop-hdds/docs/content/feature/PrefixFSO.md:30]: Do we need to add some recommendation when should it be turned on?

> This feature is strongly recommended to be turned on when Ozone buckets are mainly used via Hadoop compatible interfaces, especially with high number of files in deep directory hierachy.

**Comment by rakeshadr** [comment]
Thanks a lot @mukul1987  and @elek for the review comments. I've addressed your comments and updated the PR.

File: hadoop-hdds/docs/content/feature/PrefixFSO.md (hunk 1 above)
Comment (rakeshadr) [reply hadoop-hdds/docs/content/feature/PrefixFSO.md:61]: Done!

File: hadoop-hdds/docs/content/feature/PrefixFSO.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+```XML
+<property>
+   <name>ozone.om.enable.filesystem.paths</name>
+   <value>true</value>
+</property>
+<property>
+   <name>ozone.om.metadata.layout</name>
+   <value>prefix</value>
+</property>
+```
```
[the remark this answers is not in the document]
Comment (rakeshadr) [reply hadoop-hdds/docs/content/feature/PrefixFSO.md:75]: Done!

**Comment by rakeshadr** [comment]
> Thanks to create this patch @prashantpogde
> 
> I really like the quick summary, I think this is the important part: explaining what are the benefits of turning of this feature and what are the consequences.
> 
> I also like the summary presentation uploaded to the parent jira. I would suggest including the main diagram in the doc, which explains the differences very well:
> 
> image
> 
> But the two other (delete / rename) pictures also can be included (IMHO) as they explain the problem very well.

@elek  I have included three images, please take a look at it. Thanks!

**Comment by adoroszlai** [comment]
@rakeshadr findbugs fix (6822b10) needs to be merged into the feature branch.

**Comment by rakeshadr** [comment]
> @rakeshadr findbugs fix ([6822b10](https://github.com/apache/ozone/commit/6822b1074ee72e008d2aa6c927bb82069477370b)) needs to be merged into the feature branch.


Thanks @elek. I have rebased the feature branch with master to include the findbug fix. I hope the build will be happy now:-)

**Comment by rakeshadr** [comment]
> @rakeshadr findbugs fix ([6822b10](https://github.com/apache/ozone/commit/6822b1074ee72e008d2aa6c927bb82069477370b)) needs to be merged into the feature branch.

Thanks @adoroszlai , I've merged it.

**Review by elek — APPROVED** [review]
+1, thanks the update @rakeshadr 

Checked with Hugo and images are linked well.

(I am not sure if DirectoryDeletingService.java is supposed to be included here...)

**Review by mukul1987 — APPROVED** [review]
+1, LGTM

**Comment by rakeshadr** [comment]
> +1, thanks the update @rakeshadr
> 
> Checked with Hugo and images are linked well.
> 
> (I am not sure if DirectoryDeletingService.java is supposed to be included here...)

Thanks @elek for the comments. Included a minor typo fix in DDS java docs.

**Comment by rakeshadr** [comment]
Thanks @mukul1987 for the reviews.

**Pull request merged**